snapcraft/plugins/v2/meson.py

```diff
@@ -0,0 +1,100 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The meson plugin is useful for building meson based parts.
+
+Meson based projects are projects that have a meson.build that drives the
+build.
+
+This plugin leverages ninja to build and install.
+
+Additionally, this plugin uses the following plugin-specific keywords:
+
+    - meson-version:
+      (string)
+      The version of meson to install from PyPI.
+      If unspecified, the latest released version of meson will be used.
+    - meson-parameters:
+      (list of strings)
+      Configure flags to pass to the build using the common meson semantics.
+
+This plugin also interprets these specific build-environment entries:
+
+    - SNAPCRAFT_MESON_BUILDTYPE:
+      (default: release)
+      Use as the default --buildtype for meson when not explicitly set in
+      meson-parameters.
+"""
+
+from typing import Any, Dict, List, Set
+
+from snapcraft.plugins.v2 import PluginV2
+
+
+class MesonPlugin(PluginV2):
+    @classmethod
+    def get_schema(cls) -> Dict[str, Any]:
+        return {
+            "$schema": "http://json-schema.org/draft-04/schema#",
+            "type": "object",
+            "additionalProperties": False,
+            "properties": {
+                "configflags": {
+                    "type": "array",
+                    "minitems": 1,
+                    "uniqueItems": True,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+                "meson-version": {"type": "string", "default": ""},
+            },
+            "required": ["source"],
+        }
+
+    def get_build_snaps(self) -> Set[str]:
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        return {
+            "ninja-build",
+            "gcc",
+            "python3-pip",
+            "python3-setuptools",
+            "python3-wheel",
+        }
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {"SNAPCRAFT_MESON_BUILDTYPE": "release"}
+
+    def get_build_commands(self) -> List[str]:
+        if self.options.meson_version:
+            meson_package = f"meson=={self.options.meson_version}"
+        else:
+            meson_package = "meson"
+
+        meson_cmd = ["meson"]
+        if self.options.meson_parameters:
+            meson_cmd.append(" ".join(self.options.meson_parameters))
+        if not any(c.startswith("--buildtype=") for c in self.options.meson_parameters):
+            meson_cmd.append("--buildtype=${SNAPCRAFT_MESON_BUILDTYPE}")
+        meson_cmd.append(".")
+
+        return [
+            f"python3 -m pip install -U {meson_package}",
+            " ".join(meson_cmd),
+            "ninja",
+            'ninja install DESTDIR="${{SNAPCRAFT_PART_INSTALL}}"',
```

Comment:
not a format string, the double brackets aren't required

Reply:
thanks, I have t fixed locally, but was wrapping up my nodejs->npm plugin :)

Reply:
or maybe not :sweat_smile:

Reply:
done